### ISSUE TITLE
code splitting / chunking of cache.json files

### ISSUE THREAD

**Issue by thescientist13** [ProjectEvergreen/greenwood#317]
## Type of Change
- [ ] New Feature Request
- [ ] Documentation / Website
- [x] Improvement / Suggestion
- [ ] Bug
- [ ] Other (please clarify below)

## Summary
While [code splitting](#265 ) helps with chunking horizontally across routes, for a full graph with hundreds or thousands of pages, a call to the graph would result in huge _cache.json_ file.

We need to think about how to split this up, so _cache.json_ file can be chunked too.

**Comment by thescientist13** [comment]
@hutchgrant 
I think it would be good to edit the description here with what you had shared in Slack.  Added this issue to project 5.

**Comment by hutchgrant** [comment]
Yes I agree. This is a concern we will need to address or scaling larger than 300 pages or using too much data on <100 could slow page load times.

**Comment by thescientist13** [comment]
I wonder if something like [**unified**](https://unifiedjs.com/) would be helpful with either this or #305 ?

**Comment by thescientist13** [comment]
[Tweet from Dan Abramov](https://twitter.com/dan_abramov/status/1251927220030976000)

> _But about a typical app architecture. Such as a fully client side app. **Or a server-rendered app with a poor approach (like a giant JSON blob in HTML blocking interactions)**_.

**Comment by thescientist13** [comment]
Docs page transfers 2.6MB of JSON from _cache.json_ 😞 
<img width="1672" alt="Screen Shot 2020-04-24 at 4 34 53 PM" src="https://user-images.githubusercontent.com/895923/80254670-980dd780-8649-11ea-984c-5af41bf596fa.png">

**Comment by hutchgrant** [comment]
The deepmerge within serialize.js is where we're getting this large cache from.  There are 2 queries in the app-template (one isn't necessary), one in the header, and one in the shelf. That's how you have so many requests being made. Also I don't know if our hydration logic is correct either.

In any event, the deepmerge approach can be tweaked so that maybe we dont include 8 pages worth of queries into a single file?

**Comment by hutchgrant** [comment]
```js
    const response = await Promise.all([
      await client.query({
        query: ConfigQuery
      }),
      await client.query({
        query: GraphQuery
      })
    ]);

```

That graph query is massive and when each page is serialized that will be added to every single cache.  It's only needed once on initial page load for SPA.

**Comment by thescientist13** [comment]
Some thoughts on quick wins here
1. Fix [overfetching](https://github.com/ProjectEvergreen/greenwood/issues/317#issuecomment-619227081)
1. Reduce [duplication of query results within a single _cache.json_](https://github.com/ProjectEvergreen/greenwood/blob/master/packages/cli/src/lifecycles/serialize.js#L21) file (e.g. _about/cache.json_)

For something like _docs/cache.json_, this could bring the size down to 35k!  We should also try and add a "budget" spec for this to make sure the size can't explode again without it failing at least.

---

Later down the road, we should look to reduce duplication of query results across _all cache.json_ files, e.g. (_about/cache.json_, _docs/cache.json_)

**Comment by thescientist13** [comment]
Moving the conversation to https://github.com/ProjectEvergreen/greenwood/issues/345 for the performance and optimizations solutions.  The original intent of this issue is too far ahead of our immediate needs right now so don't want to conflate these things too much.

**Comment by thescientist13** [comment]
Another thought with this, like for the graph, is to maybe just be able to filter it per route, like we do with menu?  This might help cut down a lot on data returned per _index.html_, but then you might need to save _cache.json_ per route and sub route, instead.

**Comment by thescientist13** [comment]
Resolved as part of #383 / #371
